src/TTImp/Elab/Ambiguity.idr

```diff
@@ -74,12 +74,8 @@ expandAmbigName mode nest env orig args (IVar fc x) exp
     visible (n, i, def)
         = do let NS ns x = fullname def
                  | _ => pure True
-             if !(isVisible ns)
-                then if visibleInAny (!getNS :: !getNestedNS) (NS ns x)
-                                     (visibility def)
-                        then pure True
-                        else pure False
-                else pure False
+             pure $ !(isVisible ns) &&
+                    visibleInAny (!getNS :: !getNestedNS) (NS ns x) (visibility def)
```

Comment:
I think that this one does not actually have the same meaning
as the `!`-prefixed expressions are effectful computations.
This gets desugared to:

```idris
a <- isVisible ns
b <- getNS
c <- getNestedNS
pure (a && visibleAny (b :: c) (NS n x) (visibility def))
```

instead of

```idris
a <- isVisible ns
if a then do
  b <- getNS
  c <- getNestedNS
  pure (visibleAny (b :: c) (NS n x) (visibility def))
else pure False
```

Reply:
Here is my reasoning for the change. To start off, we have

```idris
             if !(isVisible ns)
                then if visibleInAny (!getNS :: !getNestedNS) (NS ns x) (visibility def)
                        then pure True
                        else pure False
                else pure False
```

There are three exit points from this `if` expression, all of which are `pure`. Hoisting that out:

```idris
             pure $ if !(isVisible ns)
                       then if visibleInAny (!getNS :: !getNestedNS) (NS ns x) (visibility def)
                            then True
                            else False
                       else False
```

`visibleInAny` just returns `Bool`, so that whole interior `if` expression can be reduced:

```idris
             pure $ if !(isVisible ns)
                       then visibleInAny (!getNS :: !getNestedNS) (NS ns x) (visibility def)
                       else False
```

At this point the `if` expression should desugar to:

```idris
a <- isVisible ns
if a
	then visibleInAny (!getNS :: !getNestedNS) (NS ns x) (visibility def)
	else False
```

This is just to say that `a` and the `visibleInAny` call must both be true, so join them as a conjunction:

```idris
a <- isVisible ns
a && visibleInAny (!getNS :: !getNestedNS) (NS ns x) (visibility def)
```

Which in turn desugars to:

```idris
a <- isVisible ns
a && do b <- getNS
        c <- getNestedNS
		visibleInAny (b :: c) (NS ns x) (visibility def)
```

`&&` is lazy, so those `b` and `c` assignments won't get executed unless `a` has come back true.

Or does `!` have some kind of precedence that overrules `&&`? To me that would be strange and unexpected behavior. But then, I don't understand Idris side effects very well to begin with. There is probably something important that I'm missing.

Reply:
`&&` is pure so its second argument cannot be a do block.
That's why it's all hoisted before the `pure`.

Reply:
The ! gets lifted no higher than the beginning of the block (which would be the start of the `then` and `else` branches here). iirc the reason for the `if` rather than `&&` is because the ! will get executed even if it's not needed because it's lifted outside the `&&`. If you see something of the form `if x then True else False` that's typically the most likely explanation.

Although, in this case, the operations are just lookups, so might even be cheaper than the `Delay`. So clarity is probably more important.

Reply:
For consistency with other similar uses, I'll change just the interior `if` block. As a user this behavior is surprising and confusing to me. I would have expected at least for the compiler to say something about it.